webhook/pr_review.go

```diff
@@ -0,0 +1,384 @@
+	}(r.Body)
+	if err != nil {
+		http.Error(w, "Error reading request body", http.StatusInternalServerError)
+		return
+	}
+
+	var event PullRequestEvent
+	err = json.Unmarshal(body, &event)
+	if err != nil {
+		http.Error(w, "Error unmarshaling request body", http.StatusInternalServerError)
+		return
+	}
+
+	if event.Action == "opened" || event.Action == "reopened" || event.Action == "synchronize" {
+		log.Println("Pull request event received:", event.Action)
+		owner, repo := parseRepoInfo(event.Repository.FullName)
+		var files []File
+		files, err = fetchPRFiles(owner, repo, event.Number)
+		if err != nil {
+			log.Println("Error fetching PR files:", err)
+			return
+		}
+
+		commitID, err := fetchPRCommitID(owner, repo, event.Number)
+		log.Println("Commit ID:", commitID)
+		if err != nil {
+			log.Println("Error fetching PR commit ID:", err)
+			return
+		}
+		submitReview(owner, repo, event.Number, commitID, files)
+	} else {
+		log.Println("Ignored event:", event.Action)
+	}
+}
+
+func initGithubParameter() {
+	apiBaseurl = viper.GetString("git.api_base_url")
+	token = viper.GetString("git.token")
+	owner = viper.GetString("git.owner")
+	repo = viper.GetString("git.repo")
+	sleepTimeSecond = viper.GetInt("server.sleep_time_second")
+	log.Printf("apiBaseurl=%s, token=%s, owner=%s, repo=%s\n", apiBaseurl, token, owner, repo)
+	initFlag = true
+}
+
+func parseRepoInfo(fullName string) (owner, repo string) {
+	s := strings.Split(fullName, "/")
+	return s[0], s[1]
+}
+
+func fetchPRFiles(owner, repo string, prNumber int) ([]File, error) {
+	client := &http.Client{}
+
+	req, err := http.NewRequest("GET", fmt.Sprintf(fetchPRFilesURL, apiBaseurl, owner, repo, prNumber), nil)
+	if err != nil {
+		log.Println("Error creating request:", err)
+		return nil, err
+	}
+
+	req.Header.Set("Accept", "application/vnd.github+json")
+	req.Header.Set("Authorization", "token "+token)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Println("Error sending request:", err)
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusOK {
+		bodyBytes, err := io.ReadAll(resp.Body)
+		if err != nil {
+			log.Println("Error reading response body:", err)
+			return nil, err
+		}
+
+		var files []File
+		err = json.Unmarshal(bodyBytes, &files)
+		if err != nil {
+			log.Println("Error unmarshaling response body:", err)
+			return nil, err
+		}
+		return files, nil
+	} else {
+		log.Printf("Error fetching PR files: %d\n", resp.StatusCode)
+		return nil, fmt.Errorf("Error fetching PR files: %d", resp.StatusCode)
+	}
+}
+
+func fetchPRCommitID(owner, repo string, prNumber int) (string, error) {
+	client := &http.Client{}
+
+	req, err := http.NewRequest("GET", fmt.Sprintf(fetchCommitsURL, apiBaseurl, owner, repo, prNumber), nil)
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", "token "+token)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", err
```

Comment:
Overall, the code patch looks well-written and organized. However, there are a few areas that could be improved:

1. Error handling: While the code does have error handling, it could be improved by providing more descriptive error messages and handling errors with more granularity. For example, instead of returning a generic error message, it would be better to provide specific information about the type of error that occurred.

2. Security: The code is using a token for authentication, which is good. However, it would be better to use environment variables or a secure configuration file to store sensitive information like the token, instead of using viper.

3. Code optimization: The code could be optimized by reducing the number of API calls made to GitHub. For example, instead of fetching PR files and commits separately, the code could be modified to fetch both in a single API call.

4. Code readability: The code could be made more readable by using more descriptive variable and function names, and commenting code sections that may not be obvious to other developers.

5. Code testing: The code does not have any unit tests, which could make it harder to maintain and debug in the future. Adding unit tests would increase code reliability and ensure that any changes made do not break existing functionality.

Overall, the code patch looks good, but there is room for improvement in terms of error handling, security, code optimization, readability, and testing.